File: BOJ_2178.cpp
//https://www.acmicpc.net/problem/2178
#include <iostream>
using namespace std;

#define MAX ((int)100)
char map[MAX+10][MAX+10];
int visited[MAX+10][MAX+10];

int Y,X;

struct QUE{
	int y;
	int x;
	int t;
};

QUE que[MAX*MAX+10];
int wp,rp;

void push(int y, int x, int t){
	que[wp].y=y; que[wp].x=x;que[wp].t=t;
	wp++;
}
void pop(){rp++;}
QUE front(){return que[rp];}
int empty(){return wp==rp;}

int direct[4][2]={1,0,
								 -1,0,
								 0,1,
								 0,-1};

int BFS(int s, int e, int t){
	wp=rp=0;
	push(s,e,t);
	while(rp!=wp){
		QUE cur = front();pop();
		if(cur.y==(Y-1) && cur.x==(X-1)) return cur.t;
		for(int i=0;i<4;i++){
			int dy=cur.y+direct[i][0];
			int dx=cur.x+direct[i][1];
			if(dy<0||dx<0||dy>=Y||dx>=X)continue;
			if(visited[dy][dx]==1)continue;
			if(map[dy][dx]=='0')continue;
			visited[dy][dx]=1;
			push(dy,dx,cur.t+1);
		}
	}
	return -1;
}
int main() {
	cin>>Y>>X;
	for(int y=0;y<Y;y++){
		cin>>map[y];
	}
	int res = BFS(0,0,1);
	cout<<res;
	return 0;
}

File: BOJ_2178.py
#https://www.acmicpc.net/problem/2178
# -*- coding: utf-8 -*-
# UTF-8 encoding when using korean
#import sys
from queue import Queue
from sys import stdin

def BFS(que):
	direct=[[1,0],[-1,0],[0,1],[0,-1]]
	global Y,X
	que.put((0,0,1)) #y,x,t
	while (not que.empty()):
		y,x,t=que.get()		
		if y==(Y-1) and x==(X-1):
			return t
		for i in range(4):
			dy=y+direct[i][0]
			dx=x+direct[i][1]
			if dy>=Y or dy<0 or dx>=X or dx<0:
				continue
			if visited[dy][dx] == 1 : 
				continue
			# if arr[dy][dx] == 0 :
			if arr[dy][dx] == '0':
				continue 
			visited[dy][dx]=1	
			que.put((dy,dx,t+1))
	return -1
	
que=Queue()
# Y,X=map(int, stdin.readline().split())
# arr = [[0 for _ in range(X)] for _ in range(Y)]
# visited = [[0 for _ in range(X)] for _ in range(Y)]
# for i in range(Y):
#     line=stdin.readline().strip()
#     for j,v in enumerate(line):
#         arr[i][j]=int(v)
Y, X = map(int, stdin.readline().split())
arr = [stdin.readline().rstrip() for _ in range(Y)] #possible with String Type
visited = [[0]*X for _ in range(Y)]

res = BFS(que)
print(res)
